### PR TITLE
Add support for `realm.user_push` calls

### PR DESCRIPTION
**Pull request by ericmann**
Enable Realm-signed calls for user push functionality for authentication via a mobile device.

Adds four methods:
- `RealmApi::userPush()` - push with one of user_id/email/username
- `RealmApi::userPushById()` - push with user_id
- `RealmApi::userPushByEmail()` - push with email
- `RealmApi::userPushByUsername()` - push with username

To use email/username, these fields must be configured as "distinguished" fields within the Tozny administration dashboard, mapping to "tozny_email" and "tozny_username," respectively.

### CODE REVIEW

File: src/main/java/com/tozny/sdk/RealmApi.java [hunk 1]
```diff
@@ -306,6 +315,62 @@ public OTPChallenge otpChallenge(String type, String context, String destination
...
+     * One of `userId`, `email`, or `username` must be supplied.
+     *
+     * @param sessionId The session whose owner we are testing.
+     * @param userId    Optional ID of the user to authenticate
+     * @param email     Optional email of the user to authenticate
+     * @param username  Optional username of the user to authenticate
+     *
+     * @return true on success
+     */
+    public Boolean userPush(String sessionId, String userId, String email, String username) {
```
Comment (gstro) [inline src/main/java/com/tozny/sdk/RealmApi.java]: I think adding `@Nullable` could be useful for the optional args.

**Review by gstro — COMMENTED** [review]
LGTM

File: src/main/java/com/tozny/sdk/realm/methods/user_push/UserPushRequest.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+public class UserPushRequest implements ToznyApiRequest {
+
+    @JsonProperty private String method = "realm.user_push";
+    @JsonProperty private String session_id;
+    @JsonProperty private String user_id;
+    @JsonProperty private String email;
+    @JsonProperty private String username;
+
+    public UserPushRequest(String session_id, @Nullable String user_id, @Nullable String email, @Nullable String username) {
+        if (user_id == null && email == null && username == null) {
```
Comment (m4dc4p) [inline src/main/java/com/tozny/sdk/realm/methods/user_push/UserPushRequest.java:23]: What if these value are blank or all whitespace?

File: src/main/java/com/tozny/sdk/RealmApi.java (hunk 1 above)
Comment (ericmann) [reply src/main/java/com/tozny/sdk/RealmApi.java]: Done :-)

File: src/main/java/com/tozny/sdk/RealmApi.java [hunk 3]
```diff
@@ -247,10 +247,11 @@ private boolean dispatchUserExists(UserExistsRequest req) throws ToznyApiExcepti
     }
 
     private boolean dispatchUserPush(UserPushRequest request) throws ToznyApiException {
-        ToznyApiResponse<Boolean> response = protocol.<ToznyApiResponse<Boolean>>dispatch(
-                request, new TypeReference<ToznyApiResponse<Boolean>>() {});
-        String ret = response.getReturn();
-        return ret != null && ret.equals("true");
+        ToznyApiResponse<Boolean[]> response = protocol.<ToznyApiResponse<Boolean[]>>dispatch(
```
Comment (m4dc4p) [inline src/main/java/com/tozny/sdk/RealmApi.java:252]: The server really returns a one-element array?

File: src/main/java/com/tozny/sdk/RealmApi.java [hunk 4]
```diff
@@ -247,10 +247,11 @@ private boolean dispatchUserExists(UserExistsRequest req) throws ToznyApiExcepti
...
+                request, new TypeReference<ToznyApiResponse<Boolean[]>>() {});
+
+        Boolean[] results = response.getResult();
+        return results != null && results.length == 1 && results[0];
```
Comment (m4dc4p) [inline src/main/java/com/tozny/sdk/RealmApi.java]: This expression doesn't distinguish wrong results from the result of the operation.
- `results ==  null` - Server error
- `results.length != 1` - Server error
- `results[0] == null` - Server error (and missing)
- `results[0]` - Result of the operation  

Does the SDK distinguish server errors somehow? I think the right thing is to throw an exception, but I would rather keep this consistent with existing code (i.e., not have to modify all of it).

File: src/main/java/com/tozny/sdk/realm/methods/user_push/UserPushRequest.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+@JsonAutoDetect(getterVisibility=JsonAutoDetect.Visibility.NONE)
+public class UserPushRequest implements ToznyApiRequest {
+
+    @JsonProperty private String method = "realm.user_push";
+    @JsonProperty private String session_id;
+    @JsonProperty private String user_id;
+    @JsonProperty private String email;
+    @JsonProperty private String username;
+
+    public UserPushRequest(String session_id, @Nullable String user_id, @Nullable String email, @Nullable String username) {
```
Comment (m4dc4p) [inline src/main/java/com/tozny/sdk/realm/methods/user_push/UserPushRequest.java:22]: I think a Builder would be really useful for this class. Having four string arguments can be confusing and easy to screw up. Instead, make this constructor private and have a Builder that calls it. 

Also improves error handling and will make the overloads you define in a a later commit cleaner.

**Review by m4dc4p — COMMENTED** [review]
Looks good overall. I think UserPushRequest would benefit from a Builder, and it's surprising to me that the UserPushResponse is an array of Booleans rather than a single Boolean (but maybe it refers to the result of multiple operations?).

In any case, please let me know what you think.

File: src/main/java/com/tozny/sdk/realm/methods/user_push/UserPushRequest.java (hunk 5 above)
Comment (ericmann) [reply src/main/java/com/tozny/sdk/realm/methods/user_push/UserPushRequest.java:22]: Everything is built with the same pattern for some fancy dispatch/serialization work. While I agree in principal, refactoring this one call to follow a different pattern is more change than we want to introduce for just this one quick API call addition. We do have a rewrite of the SDK targeted for a future date already that will help clarify a lot of this, though, and that will be a good change for down the road.

File: src/main/java/com/tozny/sdk/realm/methods/user_push/UserPushRequest.java (hunk 2 above)
Comment (ericmann) [reply src/main/java/com/tozny/sdk/realm/methods/user_push/UserPushRequest.java:23]: If they're all blank or whitespace, the API itself will return an exception that is caught by the SDK and converted to a `ToznyApiException`.

File: src/main/java/com/tozny/sdk/RealmApi.java (hunk 3 above)
Comment (ericmann) [reply src/main/java/com/tozny/sdk/RealmApi.java:252]: Yes. The literal JSON returned from the API is either:

``` json
{"return": "ok", "status_code": 200, "results": [false], "total": 1}
```

Or

``` json
{"return": "ok", "status_code": 200, "results": [true], "total": 1}
```

We're looking for the contents of the `results` array.

File: src/main/java/com/tozny/sdk/RealmApi.java (hunk 4 above)
Comment (ericmann) [reply src/main/java/com/tozny/sdk/RealmApi.java]: If there's an exception, it's caught and cast as a `TonzyApiException`. All other returns from this method are intended to just be `false` to indicate the push did not go out. The API itself literally only returns `true` or `false` and doesn't indicate _why_ the push failed, just that it failed.

Comment (m4dc4p) [reply src/main/java/com/tozny/sdk/RealmApi.java]: Ok, you should check for `results[0] == null` though.

File: src/main/java/com/tozny/sdk/realm/methods/user_push/UserPushRequest.java (hunk 5 above)
Comment (m4dc4p) [reply src/main/java/com/tozny/sdk/realm/methods/user_push/UserPushRequest.java:22]: Makes sense to me.

File: src/main/java/com/tozny/sdk/RealmApi.java (hunk 4 above)
Comment (ericmann) [reply src/main/java/com/tozny/sdk/RealmApi.java]: Can-do

**Pull request merged**